pkg/policy/bucket-policy.go

```diff
@@ -76,12 +78,39 @@ var startsWithFunc = func(resource string, resourcePrefix string) bool {
 	return strings.HasPrefix(resource, resourcePrefix)
 }
 
-// User - canonical users list.
-type User struct {
+// _User - canonical users list.
+type _User struct {
 	AWS           set.StringSet `json:"AWS,omitempty"`
 	CanonicalUser set.StringSet `json:"CanonicalUser,omitempty"`
 }
 
+// User - canonical users list, it supports both
+// json string and json represented by _User struct
+type User _User
+
+// UnmarshalJSON is a custom json unmarshaler for Principal field
+func (u *User) UnmarshalJSON(data []byte) error {
+	// Try to marshal a json string and see if it is a star
+	var star string
+	err := json.Unmarshal(data, &star)
+	if err == nil {
+		if star != "*" {
+			return errors.New("unrecognized Principal field")
+		}
+		*u = User{AWS: set.CreateStringSet("*")}
+		return nil
+	}
+	// Try to do regular json unmarsheling
+	var _u _User
```

Comment:
You can create a struct inside UnmarshalJSON than exposing it outside.. @vadmeste

Reply:
> You can create a struct inside UnmarshalJSON than exposing it outside.. @vadmeste

@harshavardhana, if I add the struct inside UnmarshalJSON, wouldn't I be obliged to duplicate that struct outside ?

Reply:
No you will be doing it a bit differently. Will show you an example.

Reply:
> No you will be doing it a bit differently. Will show you an example.

ping @harshavardhana :)